mater/lib/src/stores/filestore.rs

```diff
@@ -45,28 +54,60 @@ where
             // this means there is no right way of knowing when the reader is fully exhausted!
             // If we need to support a case like that, we just need to track how many times
             // the reader returned 0 and break at a certain point
-            if source.read_buf(&mut buf).await? == 0 {
-                // EOF but there's still content to yield -> yield it
+            let read_bytes = source.read_buf(&mut buf).await?;
+            println!("Read {} bytes, buffer now has {} bytes", read_bytes, buf.len());
+
+            if read_bytes == 0 {
                 if buf.len() > 0 {
+                    total_chunks += 1;
+                    println!("Final chunk {}: {} bytes", total_chunks, buf.len());
                     let chunk = buf.split();
                     yield chunk.freeze();
                 }
-                break
-            } else if buf.len() >= chunk_size {
+                break;
+            }
+
+            // While we have enough for a full chunk
+            while buf.len() >= chunk_size {
+                total_chunks += 1;
+                println!("Full chunk {}: {} bytes", total_chunks, chunk_size);
                 // The buffer may have a larger capacity than chunk_size due to reserve
                 // this also means that our read may have read more bytes than we expected,
                 // thats why we check if the length if bigger than the chunk_size and if so
                 // we split the buffer to the chunk_size, then freeze and return
                 let chunk = buf.split_to(chunk_size);
                 yield chunk.freeze();
-            } // otherwise, the buffer is not full, so we don't do a thing
+            }
+            // If we reach EOF but still have a partial chunk, yield it
+            if read_bytes == 0 && buf.len() > 0 {
+                total_chunks += 1;
+                println!("Partial chunk {}: {} bytes", total_chunks, buf.len());
+                let chunk = buf.split();
+                yield chunk.freeze();
+            }
+        }
+        println!("Total chunks yielded: {}", total_chunks);
+    });
+
+    let nodes: Peekable<SendableStream> = match wrap_mode {
+        WrapMode::Raw => futures::stream::StreamExt::peekable(Box::pin(
+            stream_balanced_tree(chunker, tree_width).map(|result| {
+                result.map(|(cid, bytes)| (Cid::from(cid), bytes.to_vec()))
+            })
+        )),
+        WrapMode::UnixFS => {
+                futures::stream::StreamExt::peekable(Box::pin(
+                    stream_balanced_tree_unixfs(chunker, tree_width)
+                        .map(|res| res)
+                ))
+
         }
     };
```

Comment:
```suggestion
    let nodes: Peekable<SendableStream> = match wrap_mode {
            stream_balanced_tree(chunker, tree_width).map(|result| {
                result.map(|(cid, bytes)| (Cid::from(cid), bytes.to_vec()))
            })
            .boxed()
            .peekable(),
        WrapMode::UnixFS => {
            stream_balanced_tree_unixfs(chunker, tree_width)
            .map(|res| res)
            .boxed()
            .peekable()
        }
    };
```

Didn't try. Should work.